rectools/models/bert4rec.py

```diff
@@ -0,0 +1,578 @@
+        interactions = interactions[interactions[Columns.User].isin(users_internal)]  # todo: fast_isin
+        interactions = interactions[interactions[Columns.Item].isin(items_internal)]
+
+        # Convert to external ids
+        interactions[Columns.Item] = dataset.item_id_map.convert_to_external(interactions[Columns.Item])
+        interactions[Columns.User] = dataset.user_id_map.convert_to_external(interactions[Columns.User])
+
+        # Prepare new user id mapping
+        rec_user_id_map = IdMap.from_values(interactions[Columns.User])
+
+        # Construct dataset
+        # TODO: For now features are dropped because model doesn't support them
+        n_filtered = len(users) - rec_user_id_map.size
+        if n_filtered > 0:
+            explanation = f"""{n_filtered} target users were considered cold because of missing known items"""
+            warnings.warn(explanation)
+        filtered_interactions = Interactions.from_raw(interactions, rec_user_id_map, self.item_id_map)
+        filtered_dataset = Dataset(rec_user_id_map, self.item_id_map, filtered_interactions)
+        return filtered_dataset
+
+    def transform_dataset_i2i(self, dataset: Dataset) -> Dataset:
+        """
+        Filter out interactions and adapt id maps.
+        Final dataset will consist only of model known items during fit.
+        Final user_id_map is the same as dataset original
+        Final item_id_map is model item_id_map constructed during training
+        """
+        # TODO: optimize by filtering in internal ids
+        interactions = dataset.get_raw_interactions()
+        interactions = interactions[interactions[Columns.Item].isin(self.get_known_item_ids())]
+        filtered_interactions = Interactions.from_raw(interactions, dataset.user_id_map, self.item_id_map)
+        filtered_dataset = Dataset(dataset.user_id_map, self.item_id_map, filtered_interactions)
+        return filtered_dataset
+
+    def _collate_fn_recommend(self, batch: List[Tuple[List[int], List[float]]]) -> torch.LongTensor:
+        """Right truncation, left padding to session_max_len"""
+        x = np.zeros((len(batch), self.session_max_len))
+        for i, (ses, _) in enumerate(batch):
+            session = ses.copy()
+            session = session + [1]
+            x[i, -len(ses) :] = ses[-self.session_max_len :]
+        return torch.LongTensor(x)
+
+    def get_dataloader_recommend(self, dataset: Dataset) -> DataLoader:
+        """TODO"""
+        sequence_dataset = SequenceDataset.from_interactions(dataset.interactions.df)
+        recommend_dataloader = DataLoader(
+            sequence_dataset,
+            batch_size=self.batch_size,
+            collate_fn=self._collate_fn_recommend,
+            num_workers=self.dataloader_num_workers,
+            shuffle=False,
+        )
+        return recommend_dataloader
+
+
+class PointWiseFeedForward(nn.Module):
```

Comment:
Now about the feedforward differences.
There are a few:
1. The order of dropout and activation. For RELU it doesn't have any difference in the result. So let's change the order for SASRec variant and it will be the same as BERT4Rec.
2. Activation function. Let's just add and argument to PointWiseFeedForward __init__ which activation to use
3. Placing second dropout inside PointWiseFeedForward.forward or moving it to TransformerLayers.forward. We can move them to TransformerLayers in SASRecTransformerLayers so that it would be more like the classic picture of transformers architecture.

As far as I can see after all of this we will have one PointWiseFeedForward class for both models. And it's good.
Please make this change as a separate commit and check SASRec metrics.

Reply:
fixed